### PR TITLE
feat: add isLatLong, isLatitude, isLongtitude validators

### PR DESCRIPTION
**Pull request by rubiin**
added isLatLong() which is used to check whether a given set of points is latitude and longitude or not. The validation is taken from validator.js

### CODE REVIEW

File: src/validation/Validator.ts [hunk 1]
```diff
@@ -332,6 +334,17 @@ export class Validator {
         return value instanceof Boolean || typeof value === "boolean";
     }
 
+    isLatLong(value: any): boolean {
```
Comment (vlapo) [inline src/validation/Validator.ts]: Why did you copy/paste solution from validator.js? We have instance of validator.js already created. Use `this.validatorJs.isLatLong(...)`

Comment (rubiin) [reply src/validation/Validator.ts]: i was thinking of doing that but wanted the patch to look original. I can fix that if you like

Comment (vlapo) [reply src/validation/Validator.ts]: It is not a patch. We use validatorjs internally all the time.

Comment (rubiin) [reply src/validation/Validator.ts]: okay then shall i update the code or let it be

Comment (vlapo) [reply src/validation/Validator.ts]: Update. Use validator.js function. Check e.g. `isDivisibleBy` where we use validator.js function too.

Comment (rubiin) [reply src/validation/Validator.ts]: ok on it

Comment (rubiin) [reply src/validation/Validator.ts]: @vlapo fixed it

**Comment by vlapo** [comment]
I was thinking about this and maybe we should also add `@IsLatitude` and `@IsLongtitude` validators. I am pretty sure there are developers which want to separate lat and long and use object representation instead of "lat,long" string. We can use `validatorJs.isLatLong()` e.g. `validatorJs.isLatLong(lat+',0)`

Also please could you ad default error messages for all new validators.

**Comment by rubiin** [comment]
Okay sure. About the islatitude and islongitude, I can work it in another or starting tomorrow

**Comment by rubiin** [comment]
@vlapo added the default error message. sorry about that

**Comment by vlapo** [comment]
Please add it in this PR and thanks for your contributions.

**Comment by rubiin** [comment]
Yeah I have updated it there

**Comment by rubiin** [comment]
@vlapo I have done it please check

**Comment by vlapo** [comment]
What have you done? I do not see `@IsLatitude` and `@IsLongtitude` validators in your changes.

**Comment by rubiin** [comment]
added the default message for the pr above. I was thinking of separating the islongitude and islatitude in another PR

**Comment by vlapo** [comment]
No, please do it in this PR.

**Comment by rubiin** [comment]
okay sure. I will ping you after that

**Comment by rubiin** [comment]
@vlapo  its complete

File: src/validation/Validator.ts [hunk 2]
```diff
@@ -332,6 +341,30 @@ export class Validator {
         return value instanceof Boolean || typeof value === "boolean";
     }
 
+
+/**
+* Checks if a given value is a latitude.
+*/
+    isLatLong(value: any): boolean {
+
+        return validator.isLatLong(value);
```
Comment (vlapo) [inline src/validation/Validator.ts]: Use `this.validatorJs` not a global `validator`.

Comment (vlapo) [reply src/validation/Validator.ts]: Also please check if input is string.

File: src/validation/Validator.ts [hunk 3]
```diff
@@ -332,6 +341,30 @@ export class Validator {
...
+    }
+
+    isLatitude(value: any): boolean {
+        const lat = /^\(?[+-]?(90(\.0+)?|[1-8]?\d(\.\d+)?)$/;
+   
+        return lat.test(value);
```
Comment (vlapo) [inline src/validation/Validator.ts]: As I said before use `validator.js` not regex extracted from `validator.js` library. In this case:
```ts
return (typeof value === "number" || this.isString(value)) && this.isLatLong(`${value},0`);
```

File: src/validation/Validator.ts [hunk 4]
```diff
@@ -332,6 +341,30 @@ export class Validator {
...
+    }
+
+/**
+* Checks if a given value is a longitude.
+*/
+
+    isLongitude(value: any): boolean {
+        const long = /^\s?[+-]?(180(\.0+)?|1[0-7]\d(\.\d+)?|\d{1,2}(\.\d+)?)\)?$/;
+        return long.test(value);
```
Comment (vlapo) [inline src/validation/Validator.ts]: Same as before:
```ts
return (typeof value === "number" || this.isString(value)) && this.isLatLong(`0,${value}`);
```

File: src/validation/Validator.ts (hunk 3 above)
Comment (rubiin) [reply src/validation/Validator.ts]: validatorjs doesnot have a single islatitude and islongitude , so we have to use regex extracted from that

Comment (vlapo) [reply src/validation/Validator.ts]: Sure thats why we will use ```this.isLatLong(`${value},0`)```.

Comment (rubiin) [reply src/validation/Validator.ts]: oh okay

File: src/validation/Validator.ts (hunk 4 above)
Comment (rubiin) [reply src/validation/Validator.ts]: @vlapo fixed now

**Comment by rubiin** [comment]
I have fixed all the issues that you have mentioned @vlapo

File: README.md [hunk 5]
```diff
@@ -470,7 +470,7 @@ import { validate } from 'class-validator';
 
 class MyClass {
 	@MinLength(32, {
-		message: "EIC code must be at least 32 charatcers",
+		message: "EIC code must be at least 32 characters",
```
Comment (vlapo) [inline README.md]: This is already fixed in actual code base. Please do not edit lines out of scope of this PR.

File: README.md [hunk 6]
```diff
@@ -922,7 +922,10 @@ validator.isInstance(value, target); // Checks value is an instance of the targe
 | `@IsISIN()`                                     | Checks if the string is an ISIN (stock/security identifier).                                                                     |
 | `@IsISO8601()`                                  | Checks if the string is a valid ISO 8601 date.                                                                                   |
 | `@IsJSON()`                                     | Checks if the string is valid JSON.                                                                                              |
-| `@IsLowercase()`                                | Checks if the string is lowercase.                                                                                               |
+| `@IsLowercase()`                                | Checks if the string is lowercase.
```
Comment (vlapo) [inline README.md]: Formatting. Please do not edit lines out of scope of this PR.

File: README.md [hunk 7]
```diff
@@ -922,7 +922,10 @@ validator.isInstance(value, target); // Checks value is an instance of the targe
 | `@IsISIN()`                                     | Checks if the string is an ISIN (stock/security identifier).                                                                     |
 | `@IsISO8601()`                                  | Checks if the string is a valid ISO 8601 date.                                                                                   |
 | `@IsJSON()`                                     | Checks if the string is valid JSON.                                                                                              |
-| `@IsLowercase()`                                | Checks if the string is lowercase.                                                                                               |
+| `@IsLowercase()`                                | Checks if the string is lowercase.  
+| `@IsLatLong()`                                  | check if the string is a valid latitude-longitude coordinate in the format lat,long 
+| `@IsLatitude()`                                  | check if the string is a valid latitude coordinate
+| `@IsLongitude()`                                  | check if the string is a valid longitude coordinate
```
Comment (vlapo) [inline README.md]: 1. string or number - `check if the string or number is a valid latitude/longitude coordinate`

2. Again formatting. There is more spaces than is necessary. Should be:
```md
| `@IsLatLong()`                                  | check if the string is a valid latitude-longitude coordinate in the format lat,long
| `@IsLatitude()`                                 | check if the string is a valid latitude coordinate
| `@IsLongitude()`                                | check if the string is a valid longitude coordinate
```

File: src/validation/ValidationTypes.ts [hunk 8]
```diff
@@ -232,6 +235,13 @@ export class ValidationTypes {
                 return eachPrefix + "$property must be a valid ISO31661 Alpha2 code";
             case this.IS_ISO31661_ALPHA_3:
                 return eachPrefix + "$property must be a valid ISO31661 Alpha3 code";
+            case this.IS_LATLONG:
+                return eachPrefix + "$property must be a latitude,longitude string";
+
+            case this.IS_LATITUDE:
+                return eachPrefix + "$property must be a latitude string";
```
Comment (vlapo) [inline src/validation/ValidationTypes.ts]: string or number

File: src/validation/ValidationTypes.ts [hunk 9]
```diff
@@ -232,6 +235,13 @@ export class ValidationTypes {
...
+            case this.IS_LONGITUDE:
+                return eachPrefix + "$property must be a longitude string";
```
Comment (vlapo) [inline src/validation/ValidationTypes.ts]: string or number

File: test/functional/validation-functions-and-decorators.spec.ts [hunk 10]
```diff
@@ -442,6 +445,84 @@ describe("IsBoolean", function() {
...
+
+
+
+
+
+
+
+
+
+
```
Comment (vlapo) [inline test/functional/validation-functions-and-decorators.spec.ts]: What is this empty space?

File: test/functional/validation-functions-and-decorators.spec.ts [hunk 11]
```diff
@@ -442,6 +445,84 @@ describe("IsBoolean", function() {
...
+    it("should fail if validator.validate said that its invalid", function (done) {
+        checkInvalidValues(new MyClass(), invalidValues, done);
+    });
+
+});
+
+describe("IsLongitude", function () {
+
+    const validValues = ["85.3446311", "85.2100893"];
+    const invalidValues = ["853446311", "as.as12"];
```
Comment (vlapo) [inline test/functional/validation-functions-and-decorators.spec.ts]: Add also valid/invalid number values.

File: test/functional/validation-functions-and-decorators.spec.ts [hunk 12]
```diff
@@ -442,6 +445,84 @@ describe("IsBoolean", function() {
...
+
+    it("should fail if validator.validate said that its invalid", function (done) {
+        checkInvalidValues(new MyClass(), invalidValues, done);
+    }); 
+
+});
+describe("IsLatitude", function () {
+
+    const validValues = ["27.6945311", "27.675509"];
+    const invalidValues = ["276945311", "asas"];
```
Comment (vlapo) [inline test/functional/validation-functions-and-decorators.spec.ts]: Add also valid/invalid number values.

File: src/validation/Validator.ts [hunk 13]
```diff
@@ -332,6 +341,28 @@ export class Validator {
         return value instanceof Boolean || typeof value === "boolean";
     }
 
+
+/**
+* Checks if a given value is a latitude.
+*/
```
Comment (vlapo) [inline src/validation/Validator.ts]: Formatting - Indent

File: src/validation/Validator.ts [hunk 14]
```diff
@@ -332,6 +341,28 @@ export class Validator {
...
+        return this.validatorJs.isLatLong(value);
+    }
+
+    isLatitude(value: any): boolean {
+        return (typeof value === "number" || this.isString(value)) && this.isLatLong(`0,${value}`);
+    }
+
+/**
+* Checks if a given value is a longitude.
+*/
```
Comment (vlapo) [inline src/validation/Validator.ts]: Formatting - Indent

File: src/validation/ValidationTypes.ts (hunk 8 above)
Comment (vlapo) [reply src/validation/ValidationTypes.ts]: This is for IS_LATITUDE and for IS_LONGTITUDE - we allow string and number

File: src/validation/ValidationTypes.ts (hunk 9 above)
Comment (vlapo) [reply src/validation/ValidationTypes.ts]: This is for IS_LATITUDE and for IS_LONGTITUDE - we allow string and number

File: README.md (hunk 5 above)
Comment (vlapo) [reply README.md]: Please revert this.

File: README.md (hunk 6 above)
Comment (vlapo) [reply README.md]: This part is still mess up. But clearly you do not know what to do with it. I will fix it later.

File: src/validation/Validator.ts [hunk 15]
```diff
@@ -332,6 +341,34 @@ export class Validator {
...
+    isLatLong(value: any): boolean {
+
+        return this.validatorJs.isLatLong(value);
+    }
+
+    /**
+    * Checks if a given value is a latitude.
+    */
+
+
```
Comment (vlapo) [inline src/validation/Validator.ts]: Extra lines between comment and function - why?

File: src/validation/Validator.ts [hunk 16]
```diff
@@ -332,6 +341,34 @@ export class Validator {
...
+    isLatitude(value: any): boolean {
+        return (typeof value === "number" || this.isString(value)) && this.isLatLong(`0,${value}`);
+    }
+
+    /**
+    * Checks if a given value is a longitude.
+    */
+
```
Comment (vlapo) [inline src/validation/Validator.ts]: Extra lines between comment and function - why?

File: src/validation/Validator.ts [hunk 17]
```diff
@@ -332,6 +341,34 @@ export class Validator {
...
+    isLongitude(value: any): boolean {
+        return (typeof value === "number" || this.isString(value)) && this.isLatLong(`${value},0`);
+
```
Comment (vlapo) [inline src/validation/Validator.ts]: Another extra line - why?

File: README.md (hunk 5 above)
Comment (vlapo) [reply README.md]: @rubiin

File: src/validation/Validator.ts (hunk 15 above)
Comment (rubiin) [reply src/validation/Validator.ts]: just wanted to make it look clear on my ide

Comment (vlapo) [reply src/validation/Validator.ts]: Sorry but extra empty lines between comment and function related directly to comment do not look clearly. It is pointless :) Check another code in `class-validator` and try to use same formatting as we use before in our code base.

**Comment by rubiin** [comment]
my editor formated those so i rerolled and back to the previous commit and only formated my code

**Comment by rubiin** [comment]
@vlapo fixed

**Comment by yogevlahyani** [comment]
When this version is going to be published to npm?

**Comment by vlapo** [comment]
Hope this weekend :)

**Comment by DrNightmare** [comment]
Hi @vlapo and @rubiin
https://github.com/typestack/class-validator/commit/3fd15c4cc9f53a14b7a8d14aecff4628969ecc26#diff-288d846ec06621fddacbd61e08486017R284

The `IsLatitude` function is using the IS_LONGITUDE ValidationType, and vice versa.

Because of this, the `IsLatitude()` check returns the following (incorrect) error message:

"$property must be a longitude string or number"

**Pull request merged**